aten/src/ATen/native/vulkan/api/Context.cpp

```diff
@@ -0,0 +1,281 @@
+        VK_DEBUG_REPORT_WARNING_BIT_EXT |
+        VK_DEBUG_REPORT_PERFORMANCE_WARNING_BIT_EXT |
+        VK_DEBUG_REPORT_ERROR_BIT_EXT |
+        VK_DEBUG_REPORT_DEBUG_BIT_EXT,
+      debug_report_callback_fn,
+      nullptr,
+    };
+
+    const auto vkCreateDebugReportCallbackEXT =
+        (PFN_vkCreateDebugReportCallbackEXT)vkGetInstanceProcAddr(
+            instance, "vkCreateDebugReportCallbackEXT");
+
+    TORCH_CHECK(
+        vkCreateDebugReportCallbackEXT,
+        "Could not load vkCreateDebugReportCallbackEXT");
+
+    VkDebugReportCallbackEXT debug_report_callback{};
+    VK_CHECK(vkCreateDebugReportCallbackEXT(
+        instance,
+        &debugReportCallbackCreateInfo,
+        nullptr,
+        &debug_report_callback));
+  }
+
+  return instance;
+}
+
+VkPhysicalDevice acquire_physical_device(const VkInstance instance) {
+  uint32_t device_count = 0;
+  VK_CHECK(vkEnumeratePhysicalDevices(instance, &device_count, nullptr));
+  TORCH_CHECK(device_count > 0, "Vulkan: Could not find a device with vulkan support!");
+
+  std::vector<VkPhysicalDevice> devices(device_count);
+  VK_CHECK(vkEnumeratePhysicalDevices(instance, &device_count, devices.data()));
+
+  return devices[0];
+}
+
+VkPhysicalDeviceLimits query_physical_device_physical_device_limits(
+    const VkPhysicalDevice physical_device) {
+  VkPhysicalDeviceProperties physical_device_properties{};
+  vkGetPhysicalDeviceProperties(physical_device, &physical_device_properties);
+  return physical_device_properties.limits;
+}
+
+uint32_t query_compute_queue_family_index(const VkPhysicalDevice physical_device) {
+  uint32_t queue_family_count = 0;
+
+  vkGetPhysicalDeviceQueueFamilyProperties(
+      physical_device, &queue_family_count, nullptr);
+
+  TORCH_CHECK(
+      queue_family_count > 0, "Vulkan: Invalid number of queue families!");
+
+  std::vector<VkQueueFamilyProperties> queue_families_properties(
+    queue_family_count);
+
+  vkGetPhysicalDeviceQueueFamilyProperties(
+      physical_device, &queue_family_count, queue_families_properties.data());
+
+  for (uint32_t i = 0; i < queue_families_properties.size(); ++i) {
+    const VkQueueFamilyProperties& properties = queue_families_properties[i];
+    if (properties.queueCount > 0 && (properties.queueFlags & VK_QUEUE_COMPUTE_BIT)) {
+      return i;
+    }
+  }
+
+  TORCH_CHECK(
+      false, "Vulkan: Could not find a queue family that supports operations!");
+}
+
+VkDevice create_device(
+    const VkPhysicalDevice physical_device,
+    const uint32_t compute_queue_family_index) {
+  const float queue_priorities = 1.0f;
+  const VkDeviceQueueCreateInfo device_queue_create_info{
+    VK_STRUCTURE_TYPE_DEVICE_QUEUE_CREATE_INFO,
+    nullptr,
+    0u,
+    compute_queue_family_index,
+    1u,
+    &queue_priorities,
+  };
+
+  const VkDeviceCreateInfo device_create_info{
+    VK_STRUCTURE_TYPE_DEVICE_CREATE_INFO,
+    nullptr,
+    0u,
+    1u,
+    &device_queue_create_info,
+    0u,
+    nullptr,
+    0u,
+    nullptr,
```

Comment:
By the way, instance layers (extensions) and device layers (extensions) are different.  I suspect this might be the reason why we have a segfault at shutdown.  I have not tested this hypothesis, but validation layers are instance layers (hence passing null here.)